dex/src/main/scala/com/wavesplatform/dex/SpendableBalancesActor.scala

```diff
@@ -0,0 +1,98 @@
+package com.wavesplatform.dex
+
+import akka.actor.{Actor, ActorRef}
+import akka.pattern.pipe
+import cats.instances.long.catsKernelStdGroupForLong
+import cats.syntax.group.catsSyntaxGroup
+import com.wavesplatform.dex.domain.account.Address
+import com.wavesplatform.dex.domain.asset.Asset
+import com.wavesplatform.dex.fp.MapImplicits.cleaningGroup
+
+import scala.concurrent.Future
+
+class SpendableBalancesActor(spendableBalances: (Address, Set[Asset]) => Future[Map[Asset, Long]],
+                             allAssetsSpendableBalances: Address => Future[Map[Asset, Long]],
+                             addressDirectory: ActorRef)
+    extends Actor {
+
+  import context.dispatcher
+
+  /** Keeps states of all addresses by all available assets */
+  var fullState: Map[Address, Map[Asset, Long]] = Map.empty
+
+  /** Keeps balance changes of addresses for which there is no full state yet (addresses that were not requested for snapshot) */
+  var incompleteStateChanges: Map[Address, Map[Asset, Long]] = Map.empty
+
+  def receive: Receive = {
+
+    case SpendableBalancesActor.Query.GetState(address, assets) =>
+      val maybeAddressState            = fullState.get(address).orElse(incompleteStateChanges get address)
+      val assetsMaybeBalances          = assets.map(asset => asset -> maybeAddressState.flatMap(_ get asset)).toMap
+      val (knownAssets, unknownAssets) = assetsMaybeBalances.partition { case (_, balance) => balance.isDefined }
+      val knownPreparedState           = knownAssets.collect { case (a, Some(b)) => a -> b }
+
+      if (unknownAssets.isEmpty) sender ! SpendableBalancesActor.Reply.GetState(knownPreparedState)
+      else
+        spendableBalances(address, unknownAssets.keySet)
+          .map(stateFromNode => SpendableBalancesActor.NodeBalanceRequestRoundtrip(address, knownPreparedState, stateFromNode))
+          .pipeTo(self)(sender)
+
+    case SpendableBalancesActor.NodeBalanceRequestRoundtrip(address, knownState, stateFromNode) =>
+      if (fullState contains address) fullState = fullState.updated(address, fullState.getOrElse(address, Map.empty) ++ stateFromNode)
+      else incompleteStateChanges = incompleteStateChanges.updated(address, incompleteStateChanges.getOrElse(address, Map.empty) ++ stateFromNode)
+      sender ! SpendableBalancesActor.Reply.GetState(stateFromNode ++ knownState)
```

Comment:
Consider lines above and IMO we should:
1. Take values for requested keys from `fullState`, If we have `fullState` for this address
2. Take values from updated `incompleteStateChanges` otherwise

Reply:
Done